File: src/python/gui/managers/hop_map_manager.py
# Relative Path: src/python/gui/managers/hop_map_manager.py
"""
Manages creation, deletion, and management of all hop nodes.
Centralizes latency handling and topological data (edges).
Decouples core logic from Qt drawing classes.
Can later support map persistence, topological presets, or procedural generation.

Thread safety is delegated to higher-level locks when calling these methods.
No direct Qt dependencies here.
"""

import random
import math
from gui.consts.gui_consts import CLIENT_NODE_COLOR, SERVER_NODE_COLOR, HOP_NODE_COLOR

class HopMapManager:
    """
    Handles conceptual node data and edges for the entire map.
    Maintains:
      - start_node_data (dict)
      - end_node_data (dict)
      - hop_nodes_data (list of dict)
      - edges_data (list of dict)
    Node dict format example:
      {
        "node_id": 0,
        "x": <float>,
        "y": <float>,
        "radius": 15,
        "color": "#abc123",
        "label": "Client",
        "delay_ms": 30
      }
    Edge dict format example:
      {
        "from_id": 0,
        "to_id": 3,
        "color": "#000000"
      }
    """

    def __init__(self):
        self.start_node_data = None
        self.end_node_data = None
        self.hop_nodes_data = []
        self.edges_data = []

        self._last_node_id = 1

        # Default range for random latencies
        self._default_min_delay = 10
        self._default_max_delay = 50

    def set_delay_range(self, min_delay: int, max_delay: int):
        """
        Adjusts the random latency range for newly created or re-initialized nodes.
        """
        if min_delay < 0:
            min_delay = 0
        if max_delay < min_delay:
            max_delay = min_delay

        self._default_min_delay = min_delay
        self._default_max_delay = max_delay

    def _get_random_delay(self) -> int:
        """
        Returns a random delay within the configured range.
        """
        if self._default_max_delay < self._default_min_delay:
            return self._default_min_delay
        return random.randint(self._default_min_delay, self._default_max_delay)

    def initialize_map(self, total_nodes: int):
        """
        Creates the default set of nodes (start, end, and hop nodes), with a grid arrangement.
        Latency is random within the configured range. If total_nodes is unchanged, skip re-init.
        """
        current_total = 0
        if self.start_node_data:
            current_total += 1
        if self.end_node_data:
            current_total += 1
        current_total += len(self.hop_nodes_data)

        if current_total == total_nodes:
            return

        if total_nodes < 2:
            total_nodes = 2

        self.start_node_data = None
        self.end_node_data = None
        self.hop_nodes_data.clear()
        self.edges_data.clear()

        width = 1000.0
        height = 600.0

        margin = 50.0
        radius = 15
        mid_y = height / 2.0

        # Start node
        self.start_node_data = {
            "node_id": 0,
            "x": margin,
            "y": mid_y,
            "radius": radius,
            "color": CLIENT_NODE_COLOR,
            "label": "Client",
            "delay_ms": self._get_random_delay()
        }

        # End node
        end_x = width - margin
        self.end_node_data = {
            "node_id": 1,
            "x": end_x,
            "y": mid_y,
            "radius": radius,
            "color": SERVER_NODE_COLOR,
            "label": "Server",
            "delay_ms": self._get_random_delay()
        }

        hop_count = total_nodes - 2
        if hop_count > 0:
            grid_left = self.start_node_data["x"] + 100.0
            grid_right = self.end_node_data["x"] - 100.0
            if grid_right < grid_left:
                grid_left = self.start_node_data["x"]
                grid_right = self.start_node_data["x"]

            row_count = int(math.ceil(math.sqrt(hop_count)))
            col_count = int(math.ceil(hop_count / row_count))

            w = max(1.0, grid_right - grid_left)
            h = max(1.0, height - 2.0 * margin)

            cell_width = w / col_count
            cell_height = h / row_count

            for i in range(hop_count):
                node_id = i + 2
                row = i // col_count
                col = i % col_count
                cx = grid_left + (col + 0.5) * cell_width
                cy = (margin) + (row + 0.5) * cell_height

                hop_data = {
                    "node_id": node_id,
                    "x": cx,
                    "y": cy,
                    "radius": radius,
                    "color": HOP_NODE_COLOR,
                    "label": f"#{node_id - 2}",
                    "delay_ms": self._get_random_delay()
                }
                self.hop_nodes_data.append(hop_data)

        self._last_node_id = max(1, total_nodes - 1)

    def recalc_positions(self, scene_width: float, scene_height: float):
        """
        Re-lays out the existing nodes to fill the horizontal space up to scene_width,
        and centers them vertically within scene_height. Does not reset latencies or IDs.
        """
        if scene_width < 10 or scene_height < 10:
            return

        margin = 50.0
        radius = 15

        if self.start_node_data:
            self.start_node_data["x"] = margin
            self.start_node_data["y"] = scene_height * 0.5
        if self.end_node_data:
            self.end_node_data["x"] = scene_width - margin
            self.end_node_data["y"] = scene_height * 0.5

        if not self.hop_nodes_data:
            return

        grid_left = (self.start_node_data["x"] + 100.0) if self.start_node_data else 100.0
        grid_right = (self.end_node_data["x"] - 100.0) if self.end_node_data else (scene_width - 100.0)
        if grid_right < grid_left:
            grid_left = margin
            grid_right = margin

        hop_count = len(self.hop_nodes_data)
        rowsf = math.sqrt(hop_count)
        row_count = int(math.ceil(rowsf))
        col_count = int(math.ceil(hop_count / row_count))

        usable_height = scene_height - 2.0 * margin
        if usable_height < 1.0:
            return
        cell_height = usable_height / row_count
        if cell_height < (2.0 * radius):
            cell_height = 2.0 * radius

        total_grid_height = row_count * cell_height
        top_offset = 0.5 * (scene_height - total_grid_height)
        if top_offset < margin:
            top_offset = margin

        w = max(1.0, (grid_right - grid_left))
        cell_width = w / col_count
        if cell_width < (2.0 * radius):
            cell_width = 2.0 * radius

        for i, hop_data in enumerate(self.hop_nodes_data):
            row = i // col_count
            col = i % col_count
            cx = grid_left + (col + 0.5) * cell_width
            cy = top_offset + (row + 0.5) * cell_height
            hop_data["x"] = cx
            hop_data["y"] = cy

    def create_default_edges(self):
        """
        Builds a default path: start -> up to 3 nearest hops -> end.
        Stores edges in self.edges_data.
        The default path color is forced transparent so it does not clutter the view.
        """
        self.edges_data.clear()
        if not self.start_node_data or not self.end_node_data:
            return

        if not self.hop_nodes_data:
            edge_transparent = {
                "from_id": self.start_node_data["node_id"],
                "to_id":   self.end_node_data["node_id"],
                "color": "#00000000"
            }
            self.edges_data.append(edge_transparent)
            return

        interior = min(3, len(self.hop_nodes_data))
        path_node_ids = [self.start_node_data["node_id"]]
        used_indices = set()

        def dist_sq(ax, ay, bx, by):
            return (ax - bx)**2 + (ay - by)**2

        current = self.start_node_data
        for _ in range(interior):
            best_idx = -1
            best_d = 1e12
            for idx, hop in enumerate(self.hop_nodes_data):
                if idx in used_indices:
                    continue
                dsq = dist_sq(current["x"], current["y"], hop["x"], hop["y"])
                if dsq < best_d:
                    best_d = dsq
                    best_idx = idx
            if best_idx >= 0:
                path_node_ids.append(self.hop_nodes_data[best_idx]["node_id"])
                used_indices.add(best_idx)
                current = self.hop_nodes_data[best_idx]

        path_node_ids.append(self.end_node_data["node_id"])
        for i in range(len(path_node_ids) - 1):
            edge_transparent = {
                "from_id": path_node_ids[i],
                "to_id":   path_node_ids[i+1],
                "color": "#00000000"
            }
            self.edges_data.append(edge_transparent)

    def export_graph_topology(self):
        """
        Returns a dictionary containing the nodes and edges in the current map.
        This is the standard full topology export (including start/end).
        """
        nodes = []
        if self.start_node_data:
            nodes.append({
                "node_id":   self.start_node_data["node_id"],
                "delay_ms":  self.start_node_data["delay_ms"]
            })
        for h in self.hop_nodes_data:
            nodes.append({
                "node_id":  h["node_id"],
                "delay_ms": h["delay_ms"]
            })
        if self.end_node_data:
            nodes.append({
                "node_id":  self.end_node_data["node_id"],
                "delay_ms": self.end_node_data["delay_ms"]
            })

        return {
            "nodes": nodes,
            "edges": list(self.edges_data)
        }

    def export_graph_topology_for_heatmap(self):
        """
        Returns only the hop nodes and edges that do not connect
        to the start or end node (node_id=0 or node_id=1).
        This excludes the Client and Server from the heatmap.
        """
        filtered_edges = []
        for e in self.edges_data:
            if e["from_id"] not in (0,1) and e["to_id"] not in (0,1):
                filtered_edges.append(e)

        nodes = []
        for h in self.hop_nodes_data:
            nodes.append({
                "node_id":  h["node_id"],
                "delay_ms": h["delay_ms"]
            })

        return {
            "nodes": nodes,
            "edges": filtered_edges
        }

    def add_hops(self, count: int):
        """
        Adds `count` new hop nodes with unique IDs, appends to self.hop_nodes_data,
        does NOT reset start/end nodes or existing edges.
        Returns the list of newly-created hop dicts.
        """
        width = 1000.0
        height = 600.0
        margin = 50.0
        radius = 15

        grid_left = (self.start_node_data["x"] if self.start_node_data else 50.0) + 100.0
        grid_right = (self.end_node_data["x"] if self.end_node_data else 950.0) - 100.0

        new_hops = []
        for _ in range(count):
            new_id = self._last_node_id + 1
            cx = grid_left
            cy = margin
            rand_delay = self._get_random_delay()
            hop_data = {
                "node_id": new_id,
                "x": cx,
                "y": cy,
                "radius": radius,
                "color": HOP_NODE_COLOR,
                "label": f"#{new_id - 2}",
                "delay_ms": rand_delay
            }
            self.hop_nodes_data.append(hop_data)
            new_hops.append(hop_data)
            self._last_node_id = new_id

        return new_hops

File: src/c/managers/hop_map_manager.c
/* Relative Path: src/c/managers/hop_map_manager.c */
/*
 * Builds and maintains a hop-based node map, assigning positions and delays.
 * Creates default edges and exports topology data in a thread-safe manner.
 * Useful for demonstration or simple graph generation within AntNet.
 *
 * Now also exposes pub_hop_map_* functions that operate on the HopMapManager
 * stored inside AntNetContext->hop_map_mgr, consolidating node/edge creation
 * exclusively in the C side.
*/

#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <math.h>
#include <time.h>

#include "../../../include/managers/hop_map_manager.h"
#include "../../../include/consts/error_codes.h"
#include "../../../include/core/backend_init.h"       /* priv_get_context_by_id */
#include "../../../include/managers/hop_map_manager.h"/* HopMapManager, NodeData, EdgeData */

/*
  Minimal reference implementation of a HopMapManager in C.
  Thread-safety ensured via the lock mutex. This code is purely for demonstration.
  Integrate random seeding and advanced placement logic as needed.
*/

/*
 * Internal helper: distance squared
 */
static float dist_sq(float ax, float ay, float bx, float by) {
    float dx = ax - bx;
    float dy = ay - by;
    return dx*dx + dy*dy;
}

HopMapManager* hop_map_manager_create() {
    HopMapManager *mgr = (HopMapManager*)malloc(sizeof(HopMapManager));
    if (!mgr) {
        return NULL;
    }

#ifndef _WIN32
    pthread_mutex_init(&mgr->lock, NULL);
#endif

    mgr->start_node = NULL;
    mgr->end_node   = NULL;
    mgr->hop_nodes  = NULL;
    mgr->hop_count  = 0;
    mgr->edges      = NULL;
    mgr->edge_count = 0;

    /* default random delay range */
    mgr->default_min_delay = 10;
    mgr->default_max_delay = 50;

    srand((unsigned int)time(NULL));

    return mgr;
}

void hop_map_manager_destroy(HopMapManager *mgr) {
    if (!mgr) return;

#ifndef _WIN32
    pthread_mutex_destroy(&mgr->lock);
#endif

    if (mgr->start_node) free(mgr->start_node);
    if (mgr->end_node)   free(mgr->end_node);
    if (mgr->hop_nodes)  free(mgr->hop_nodes);
    if (mgr->edges)      free(mgr->edges);

    free(mgr);
}

/*
 * Allows setting the random delay range for node latencies in a thread-safe manner.
 */
void hop_map_manager_set_delay_range(HopMapManager *mgr, int min_delay, int max_delay) {
    if (!mgr) return;
#ifndef _WIN32
    pthread_mutex_lock(&mgr->lock);
#endif

    mgr->default_min_delay = (min_delay < 0) ? 0 : min_delay;
    mgr->default_max_delay = (max_delay < 0) ? 0 : max_delay;
    if (mgr->default_max_delay < mgr->default_min_delay) {
        mgr->default_max_delay = mgr->default_min_delay;
    }

#ifndef _WIN32
    pthread_mutex_unlock(&mgr->lock);
#endif
}

/*
 * Returns a random delay within the configured range.
 */
static int hop_map_manager_get_random_delay(const HopMapManager *mgr) {
    int range = mgr->default_max_delay - mgr->default_min_delay + 1;
    if (range < 1) {
        return mgr->default_min_delay;
    }
    int rnd = rand() % range;
    return mgr->default_min_delay + rnd;
}

/*
 * hop_map_manager_initialize_map
 * Allocates new node arrays for start, end, and hops if total_nodes changed.
 * If total_nodes is unchanged, it returns immediately (skips re-randomizing latencies).
 */
void hop_map_manager_initialize_map(HopMapManager *mgr, int total_nodes) {
    if (!mgr) return;

#ifndef _WIN32
    pthread_mutex_lock(&mgr->lock);
#endif

    size_t current_count = 0;
    if (mgr->start_node) current_count++;
    if (mgr->end_node)   current_count++;
    current_count += mgr->hop_count;

    if ((int)current_count == total_nodes) {
#ifndef _WIN32
        pthread_mutex_unlock(&mgr->lock);
#endif
        return;
    }

    if (total_nodes < 2) {
        total_nodes = 2;
    }

    /* free old data */
    if (mgr->start_node) { free(mgr->start_node); mgr->start_node = NULL; }
    if (mgr->end_node)   { free(mgr->end_node);   mgr->end_node   = NULL; }
    if (mgr->hop_nodes)  { free(mgr->hop_nodes);  mgr->hop_nodes  = NULL; mgr->hop_count = 0; }
    if (mgr->edges)      { free(mgr->edges);      mgr->edges      = NULL; mgr->edge_count = 0; }

    mgr->start_node = (NodeData*)malloc(sizeof(NodeData));
    mgr->end_node   = (NodeData*)malloc(sizeof(NodeData));

    if (!mgr->start_node || !mgr->end_node) {
#ifndef _WIN32
        pthread_mutex_unlock(&mgr->lock);
#endif
        return;
    }

    float width  = 1000.0f;
    float height = 600.0f;
    float margin = 50.0f;
    int radius   = 15;
    float mid_y  = height / 2.0f;

    /* Start node */
    mgr->start_node->node_id  = 0;
    mgr->start_node->x        = margin;
    mgr->start_node->y        = mid_y;
    mgr->start_node->radius   = radius;
    mgr->start_node->delay_ms = hop_map_manager_get_random_delay(mgr);

    /* End node */
    mgr->end_node->node_id  = 1;
    mgr->end_node->x        = width - margin;
    mgr->end_node->y        = mid_y;
    mgr->end_node->radius   = radius;
    mgr->end_node->delay_ms = hop_map_manager_get_random_delay(mgr);

    mgr->hop_count = (size_t)(total_nodes - 2);
    if (mgr->hop_count > 0) {
        mgr->hop_nodes = (NodeData*)malloc(sizeof(NodeData) * mgr->hop_count);
        if (!mgr->hop_nodes) {
#ifndef _WIN32
            pthread_mutex_unlock(&mgr->lock);
#endif
            return;
        }
    }

    /* Lay out hops in a grid between start and end horizontally */
    if (mgr->hop_count > 0) {
        float grid_left   = mgr->start_node->x + 100.0f;
        float grid_right  = mgr->end_node->x   - 100.0f;
        if (grid_right < grid_left) {
            grid_left  = mgr->start_node->x;
            grid_right = mgr->start_node->x;
        }

        float grid_top    = 50.0f;
        float grid_bottom = height - 50.0f;
        int total_hops = (int)mgr->hop_count;

        float rowsf = sqrtf((float)total_hops);
        int row_count = (int)ceilf(rowsf);
        int col_count = (int)ceilf((float)total_hops / (float)row_count);

        float w = (grid_right  > grid_left) ? (grid_right  - grid_left) : 1.0f;
        float h = (grid_bottom > grid_top ) ? (grid_bottom - grid_top ) : 1.0f;

        float cell_width  = w / (float)col_count;
        float cell_height = h / (float)row_count;

        for (int i = 0; i < total_hops; i++) {
            mgr->hop_nodes[i].node_id  = i + 2;
            mgr->hop_nodes[i].radius   = radius;
            mgr->hop_nodes[i].delay_ms = hop_map_manager_get_random_delay(mgr);

            int row = i / col_count;
            int col = i % col_count;

            float cx = grid_left + (col + 0.5f) * cell_width;
            float cy = grid_top  + (row + 0.5f) * cell_height;

            mgr->hop_nodes[i].x = cx;
            mgr->hop_nodes[i].y = cy;
        }
    }

#ifndef _WIN32
    pthread_mutex_unlock(&mgr->lock);
#endif
}

/*
 * hop_map_manager_recalc_positions
 * This re-lays out the *existing* start_node, end_node, and hop_nodes
 * within the specified scene_width and scene_height. The function
 * does not modify node_id or delay_ms.
 */
void hop_map_manager_recalc_positions(HopMapManager *mgr,
                                      float scene_width,
                                      float scene_height)
{
    if (!mgr) return;

#ifndef _WIN32
    pthread_mutex_lock(&mgr->lock);
#endif

    if (scene_width <= 10.0f || scene_height <= 10.0f) {
#ifndef _WIN32
        pthread_mutex_unlock(&mgr->lock);
#endif
        return;
    }

    float margin = 50.0f;
    int radius = 15;

    if (mgr->start_node) {
        mgr->start_node->x = margin;
        mgr->start_node->y = scene_height * 0.5f;
    }

    if (mgr->end_node) {
        mgr->end_node->x = scene_width - margin;
        mgr->end_node->y = scene_height * 0.5f;
    }

    if (!mgr->hop_nodes || mgr->hop_count == 0) {
#ifndef _WIN32
        pthread_mutex_unlock(&mgr->lock);
#endif
        return;
    }

    float grid_left  = (mgr->start_node) ? (mgr->start_node->x + 100.0f) : 100.0f;
    float grid_right = (mgr->end_node)   ? (mgr->end_node->x   - 100.0f) : (scene_width - 100.0f);
    if (grid_right < grid_left) {
        grid_left = margin;
        grid_right = margin;
    }

    int total_hops = (int)mgr->hop_count;
    float rowsf = sqrtf((float)total_hops);
    int row_count = (int)ceilf(rowsf);
    int col_count = (int)ceilf((float)total_hops / (float)row_count);

    float usable_height = scene_height - 2.0f * margin;
    if (usable_height < 1.0f) {
#ifndef _WIN32
        pthread_mutex_unlock(&mgr->lock);
#endif
        return;
    }

    float cell_height = usable_height / (float)row_count;
    if (cell_height < (2.0f * radius)) {
        cell_height = 2.0f * radius;
    }

    float grid_total_height = row_count * cell_height;
    float top_offset = 0.5f * (scene_height - grid_total_height);
    if (top_offset < margin) {
        top_offset = margin;
    }

    float w = (grid_right > grid_left) ? (grid_right - grid_left) : 1.0f;
    if (col_count < 1) col_count = 1;
    float cell_width = w / (float)col_count;
    if (cell_width < (2.0f * radius)) {
        cell_width = 2.0f * radius;
    }

    for (int i = 0; i < total_hops; i++) {
        int row = i / col_count;
        int col = i % col_count;

        float cx = grid_left + (col + 0.5f) * cell_width;
        float cy = top_offset + (row + 0.5f) * cell_height;

        mgr->hop_nodes[i].x = cx;
        mgr->hop_nodes[i].y = cy;
    }

#ifndef _WIN32
    pthread_mutex_unlock(&mgr->lock);
#endif
}

/*
 * Creates default edges for up to 3 nearest hops, modifies mgr->edges
 */
void hop_map_manager_create_default_edges(HopMapManager *mgr) {
    if (!mgr) return;

#ifndef _WIN32
    pthread_mutex_lock(&mgr->lock);
#endif

    if (mgr->edges) {
        free(mgr->edges);
        mgr->edges = NULL;
    }
    mgr->edge_count = 0;

    if (!mgr->start_node || !mgr->end_node) {
#ifndef _WIN32
        pthread_mutex_unlock(&mgr->lock);
#endif
        return;
    }
    if (!mgr->hop_nodes || mgr->hop_count == 0) {
        mgr->edges = (EdgeData*)malloc(sizeof(EdgeData));
        if (mgr->edges) {
            mgr->edge_count = 1;
            mgr->edges[0].from_id = mgr->start_node->node_id;
            mgr->edges[0].to_id   = mgr->end_node->node_id;
        }
#ifndef _WIN32
        pthread_mutex_unlock(&mgr->lock);
#endif
        return;
    }

    size_t interior = (mgr->hop_count < 3) ? mgr->hop_count : 3;
    int *path_node_ids = (int*)malloc(sizeof(int) * (2 + interior));
    if (!path_node_ids) {
#ifndef _WIN32
        pthread_mutex_unlock(&mgr->lock);
#endif
        return;
    }

    int *hop_idx_used = (int*)calloc(mgr->hop_count, sizeof(int));
    if (!hop_idx_used) {
        free(path_node_ids);
#ifndef _WIN32
        pthread_mutex_unlock(&mgr->lock);
#endif
        return;
    }

    path_node_ids[0] = mgr->start_node->node_id;
    NodeData current = *(mgr->start_node);
    size_t path_index = 1;

    for (size_t step = 0; step < interior; step++) {
        float best_d = 1e12f;
        int best_idx = -1;
        for (size_t h = 0; h < mgr->hop_count; h++) {
            if (hop_idx_used[h]) continue;
            NodeData *cand = &mgr->hop_nodes[h];
            float dsq = dist_sq(current.x, current.y, cand->x, cand->y);
            if (dsq < best_d) {
                best_d = dsq;
                best_idx = (int)h;
            }
        }
        if (best_idx >= 0) {
            path_node_ids[path_index++] = mgr->hop_nodes[best_idx].node_id;
            hop_idx_used[best_idx] = 1;
            current = mgr->hop_nodes[best_idx];
        }
    }

    path_node_ids[path_index++] = mgr->end_node->node_id;
    size_t final_count = path_index;

    mgr->edge_count = final_count - 1;
    mgr->edges = (EdgeData*)malloc(sizeof(EdgeData) * mgr->edge_count);
    if (mgr->edges) {
        for (size_t i = 0; i < mgr->edge_count; i++) {
            mgr->edges[i].from_id = path_node_ids[i];
            mgr->edges[i].to_id   = path_node_ids[i + 1];
        }
    }

    free(path_node_ids);
    free(hop_idx_used);

#ifndef _WIN32
    pthread_mutex_unlock(&mgr->lock);
#endif
}

void hop_map_manager_export_topology(HopMapManager *mgr,
                                     NodeData *out_nodes, size_t *out_node_count,
                                     EdgeData *out_edges, size_t *out_edge_count)
{
    if (!mgr) {
        if (out_node_count) *out_node_count = 0;
        if (out_edge_count) *out_edge_count = 0;
        return;
    }

#ifndef _WIN32
    pthread_mutex_lock(&mgr->lock);
#endif

    size_t ncount = 0;
    if (mgr->start_node) ncount++;
    if (mgr->end_node)   ncount++;
    ncount += mgr->hop_count;

    size_t ecount = mgr->edge_count;

    if (out_node_count) {
        if (out_nodes && ncount > 0) {
            size_t idx = 0;
            if (mgr->start_node) {
                out_nodes[idx++] = *(mgr->start_node);
            }
            if (mgr->hop_nodes && mgr->hop_count > 0) {
                for (size_t i = 0; i < mgr->hop_count; i++) {
                    out_nodes[idx++] = mgr->hop_nodes[i];
                }
            }
            if (mgr->end_node) {
                out_nodes[idx++] = *(mgr->end_node);
            }
        }
        *out_node_count = ncount;
    }

    if (out_edge_count) {
        if (out_edges && ecount > 0 && mgr->edges) {
            memcpy(out_edges, mgr->edges, sizeof(EdgeData) * ecount);
        }
        *out_edge_count = ecount;
    }

#ifndef _WIN32
    pthread_mutex_unlock(&mgr->lock);
#endif
}

/* ------------------------------------------------------------------
 *                 New public 'pub_' wrappers
 * ------------------------------------------------------------------
 *
 * Each function locks the same AntNetContext->lock used by the solver,
 * then calls the HopMapManager. This ensures thread safety across modules.
*/

/*
 * pub_hop_map_set_delay_range
 * Applies the given min/max to the context's HopMapManager in a thread-safe manner.
 */
int pub_hop_map_set_delay_range(int context_id, int min_d, int max_d)
{
    AntNetContext* ctx = priv_get_context_by_id(context_id);
    if (!ctx) {
        return ERR_INVALID_CONTEXT;
    }
#ifndef _WIN32
    pthread_mutex_lock(&ctx->lock);
#endif
    if (!ctx->hop_map_mgr) {
        /* Create if missing. This is optional logic. */
        ctx->hop_map_mgr = hop_map_manager_create();
        if (!ctx->hop_map_mgr) {
#ifndef _WIN32
            pthread_mutex_unlock(&ctx->lock);
#endif
            return ERR_MEMORY_ALLOCATION;
        }
    }
    /* HopMapManager usage is also locked internally, but we hold this outer
     * lock to keep everything consistent if solver or other calls happen. */
    hop_map_manager_set_delay_range(ctx->hop_map_mgr, min_d, max_d);

#ifndef _WIN32
    pthread_mutex_unlock(&ctx->lock);
#endif
    return ERR_SUCCESS;
}

/*
 * pub_hop_map_initialize
 * Initializes or re-initializes the HopMapManager with total_nodes.
 */
int pub_hop_map_initialize(int context_id, int total_nodes)
{
    AntNetContext* ctx = priv_get_context_by_id(context_id);
    if (!ctx) {
        return ERR_INVALID_CONTEXT;
    }

#ifndef _WIN32
    pthread_mutex_lock(&ctx->lock);
#endif
    if (!ctx->hop_map_mgr) {
        ctx->hop_map_mgr = hop_map_manager_create();
        if (!ctx->hop_map_mgr) {
#ifndef _WIN32
            pthread_mutex_unlock(&ctx->lock);
#endif
            return ERR_MEMORY_ALLOCATION;
        }
    }
#ifndef _WIN32
    pthread_mutex_unlock(&ctx->lock);
#endif

    /* Now call the manager's initialize_map inside the same lock. */
    hop_map_manager_initialize_map(ctx->hop_map_mgr, total_nodes);

    return ERR_SUCCESS;
}

/*
 * pub_hop_map_create_default_edges
 */
int pub_hop_map_create_default_edges(int context_id)
{
    AntNetContext* ctx = priv_get_context_by_id(context_id);
    if (!ctx) {
        return ERR_INVALID_CONTEXT;
    }

#ifndef _WIN32
    pthread_mutex_lock(&ctx->lock);
#endif
    if (!ctx->hop_map_mgr) {
#ifndef _WIN32
        pthread_mutex_unlock(&ctx->lock);
#endif
        return ERR_NO_TOPOLOGY;
    }
#ifndef _WIN32
    pthread_mutex_unlock(&ctx->lock);
#endif

    hop_map_manager_create_default_edges(ctx->hop_map_mgr);
    return ERR_SUCCESS;
}

/*
 * pub_hop_map_recalc_positions
 * Re-lays out existing nodes for the new scene size.
 */
int pub_hop_map_recalc_positions(int context_id, float scene_w, float scene_h)
{
    AntNetContext* ctx = priv_get_context_by_id(context_id);
    if (!ctx) {
        return ERR_INVALID_CONTEXT;
    }

#ifndef _WIN32
    pthread_mutex_lock(&ctx->lock);
#endif
    if (!ctx->hop_map_mgr) {
#ifndef _WIN32
        pthread_mutex_unlock(&ctx->lock);
#endif
        return ERR_NO_TOPOLOGY;
    }
#ifndef _WIN32
    pthread_mutex_unlock(&ctx->lock);
#endif

    hop_map_manager_recalc_positions(ctx->hop_map_mgr, scene_w, scene_h);
    return ERR_SUCCESS;
}

/*
 * pub_hop_map_export_topology
 * Exports the current HopMapManager node+edge data to the caller.
 * The caller supplies out_nodes/out_node_count, out_edges/out_edge_count, just
 * like pub_update_topology does for the solver. Returns 0 or negative on error.
 */
int pub_hop_map_export_topology(int context_id,
                                NodeData *out_nodes, int *out_node_count,
                                EdgeData *out_edges, int *out_edge_count)
{
    if (!out_node_count || !out_edge_count) {
        return ERR_INVALID_ARGS;
    }

    AntNetContext* ctx = priv_get_context_by_id(context_id);
    if (!ctx) {
        return ERR_INVALID_CONTEXT;
    }

#ifndef _WIN32
    pthread_mutex_lock(&ctx->lock);
#endif
    if (!ctx->hop_map_mgr) {
        /* 0 nodes/edges if not inited */
        if (out_node_count) *out_node_count = 0;
        if (out_edge_count) *out_edge_count = 0;
#ifndef _WIN32
        pthread_mutex_unlock(&ctx->lock);
#endif
        return ERR_NO_TOPOLOGY;
    }
#ifndef _WIN32
    pthread_mutex_unlock(&ctx->lock);
#endif

    size_t node_count_sz = 0;
    size_t edge_count_sz = 0;
    hop_map_manager_export_topology(ctx->hop_map_mgr,
                                    out_nodes, &node_count_sz,
                                    out_edges, &edge_count_sz);

    if (out_node_count) *out_node_count = (int)node_count_sz;
    if (out_edge_count) *out_edge_count = (int)edge_count_sz;

    return ERR_SUCCESS;
}
